C++17 round 2: constexpr if statements

trueqbit opened the pull request:
The following C++ language features a now a requirement:

* constexpr if statements. See corresponding entry in the [compiler support matrix](https://en.cppreference.com/w/cpp/compiler_support/17#cpp_if_constexpr_201606L).
* `[[maybe_unused]]` attribute

trueqbit committed:
Require `if constexpr` support

trueqbit committed:
`if constexpr`-ification

* Replaced `call_if_constexpr` and `static_if` constructs with `if constexpr`
* Used `if constexpr` for conditional constant expressions

trueqbit committed:
Merge branch 'upstream/fix_error_race_condition' into c++17_if-constexpr

trueqbit committed:
Enabled msvc warning about constant conditional expression

trueqbit committed:
Merge branch 'upstream/fix_error_race_condition' into upstream/c++17_if-constexpr

trueqbit committed:
Treat member initialization reorder as error for msvc as well

trueqbit committed:
appveyor: Updated vcpkg environment to 2025.03.19

trueqbit committed:
Attached a few more `maybe_unused` attributes

trueqbit committed:
appveyor: Replaced C++14 build jobs by C++23

fnc12 commented on dev/ast_iterator.h:
```diff
@@ -71,10 +70,9 @@ namespace sqlite_orm {
-                lambda,
-                polyfill::bool_constant<true>{},
-                t);
+            if constexpr (polyfill::is_invocable<L, polyfill::bool_constant<true>, const T&>::value) {
```
this PR breaks C++14 compatibility, right?

trueqbit replied:
Correct. C++17 is required with regard to the language standard. Further updates for the rest of C++17 that sqlite_orm uses is already in the pipeline.

fnc12 commented on dev/cte_column_names_collector.h:
```diff
@@ -159,8 +159,9 @@ namespace sqlite_orm {
-            if (size_t n = std::tuple_size_v<ExplicitColRefs>) {
-                if (n != columnNames.size()) {
+            constexpr size_t nExplicitColumns = std::tuple_size_v<ExplicitColRefs>;
+            if constexpr (nExplicitColumns > 0) {
```
this improves the lib a lot

fnc12 commented on dev/functional/static_magic.h:
```diff
@@ -24,58 +17,6 @@ namespace sqlite_orm {
         };
         template<class R = void>
         constexpr empty_callable_t<R> empty_callable{};
```
I think we don't this as well just like the whole file

trueqbit replied:
It's still in use for `statement_serializer<replace_t<T>>`.
It actually has nothing to do with "static magic", but is a useful variadic function that always returns a specific value. I can rename it? We already have a metafunction called `always_false` and a quoted metafunction `mpl::always`. I would choose `always_default`, which probably fits best..

fnc12 replied:
> I can rename it?

up to you

trueqbit replied:
Renamed and moved to `functional/always_default.h`

trueqbit committed:
Fixed CMake target compile options for msvc

trueqbit committed:
appveyor: Corrected CMake generator for Visual Studio 2019 worker image

trueqbit committed:
appveyor: Updated Ubuntu build worker image for C++23 builds

fnc12 commented on dev/implementations/storage_definitions.h:
```diff
@@ -22,18 +22,25 @@
+
+        template<class... DBO>
+        template<class Table, satisfies<is_table, Table>>
+        sync_schema_result storage_t<DBO...>::_sync_normal_table(const Table& table, sqlite3* db, bool preserve) {
```
maybe `sync_generic_table`

trueqbit replied:
You probably mean "general". I chose "regular".

fnc12 replied:
ok

trueqbit committed:
appveyor: Try to use Clang-18 for C++23 builds

fnc12 commented on dev/statement_serializer.h:
```diff
@@ -1787,23 +1788,22 @@ namespace sqlite_orm {
-                            ss << keyword << ' ';
-                        },
-                        sel.col);
+                        std::string keyword = sel.col;
```
do we have `std::string` copying here?

trueqbit replied:
Yes, see comment above, which didn't change.

fnc12 commented on dev/storage.h:
```diff
@@ -1211,6 +1208,9 @@ namespace sqlite_orm {
             sync_schema_result sync_table(const Table& table, sqlite3* db, bool preserve);
 
+            template<class Table, satisfies<is_table, Table> = true>
+            sync_schema_result _sync_normal_table(const Table& table, sqlite3* db, bool preserve);
```
func name starting with underscore looks very odd. I think regular `sync_normal_table` is well enough

trueqbit committed:
Corrected concepts used in unit tests.

* A concept cannot have an associated constraint, i.e. it cannot constrain a parameter using a concept recursively.
* dependent template names (e.g. methods with explicit template parameters) must be qualified with the `template` keyword.

trueqbit committed:
Renamed internal method that sync's regular tables

trueqbit committed:
Renamed `empty_callable` -> `always_default`

trueqbit committed:
Removed superfluous inclusion of _functional/cxx_core_features.h_

trueqbit committed:
Renamed _functional/static_magic.h_ -> _functional/always_default.h_

trueqbit committed:
Improved `same_or_void` with concepts

trueqbit committed:
Used `std::time` in a qualified way

trueqbit committed:
Added a unit test for `always_default`

trueqbit committed:
Addressed a few things found when compiling with VC 15.9

trueqbit committed:
Renamed type trait `is_base_of_template` -> `is_base_template_of`

trueqbit committed:
Coalesced checks for C++14 language features

trueqbit committed:
Corrected unit test for `always_default`

trueqbit committed:
Removed superfluous newlines introduced in commit #f96cf150bdb4

trueqbit committed:
Follow-up: Corrected concepts used in unit tests.

trueqbit committed:
Reordered template parameters of type trait `is_base_template_of`

trueqbit committed:
Derived class for a string literator operator template needs its own ctor

trueqbit committed:
Unit test for reverse tuple iteration

trueqbit committed:
Unlike msvc, gcc+clang require NTTPs to be copyable

trueqbit committed:
Ensure that a CTE's moniker type is non-const

Compiling unit tests with gcc C++23 revealed that a const moniker type was somehow part of a common table expression, which caused picking of a table from the tuple of DBOs to fail.

trueqbit committed:
Corrected unit test for gcc until v13.3

gcc until v13.3 had the bad habit to deduce a constant template parameter as const; remove it

trueqbit committed:
Further fallback correction for gcc until v13.3

trueqbit committed:
Handled non-copyable quoted scalar functions

trueqbit committed:
Replaced `(void)` casts by `maybe_unused` annotations where possible

fnc12 commented on dev/select_constraints.h:
```diff
@@ -270,30 +270,30 @@ namespace sqlite_orm {
         template<typename Moniker, class ExplicitCols>
         struct cte_builder {
-            ExplicitCols explicitColumns;
+            ExplicitCols _explicitColumns;
```
do you want to replace all `this->` places with `_` prefix?

trueqbit replied:
Do you mean across the entire library? That would be especially good for the internally created classes with public members. I would save that for another day, is that okay?

fnc12 replied:
Yes I mean across the entire lib. I am asking in terms of your global plans actually cause I am a fan if `this->` and removing it triggers me a bit. On another hand I am ok with the leading underscore. But combining both looks ugly so I am just asking.

> I would save that for another day, is that okay?

Not a problem at all.

trueqbit replied:
I am not a fan of `this->` at all, it makes things much harder to read and is usually not necessary from a syntax perspective :)
But I am OK with either style.

So we should get together one day and find a guideline for the coding style that everyone follows.
By the way, I was not the one who introduced the leading underscore. For structs as public interfaces (even if they are from the internal namespace) the leading underscore is important, as you already mentioned. Firstly because of sorting in an IDE, and secondly to easily distinguish between public and private members.

trueqbit committed:
Corrected "Macintosh" newlines

trueqbit committed:
Updated a comment on non-copyable quoted functions

trueqbit committed:
C++17 mentioned as a basic requirement in the readme
